src/tlog/verify.ts

```diff
@@ -0,0 +1,169 @@
+export function verifyTLogSET(
+  bundle: Bundle,
+  tlogKeys: Record<string, KeyObject>
+): void {
+  bundle.verificationData?.tlogEntries.forEach((entry, index) => {
+    // Re-create the original Rekor verification payload
+    const payload = toVerificationPayload(bundle, index);
+
+    // Canonicalize the payload and turn into a buffer for verification
+    const data = Buffer.from(json.canonicalize(payload), 'utf8');
+
+    // Find the public key for the transaction log which generated the SET
+    const publicKey = tlogKeys[payload.logID];
+    if (!publicKey) {
+      throw new Error('no key found for logID: ' + payload.logID);
+    }
+
+    // Extract the SET from the tlog entry
+    const signature = entry.inclusionPromise?.signedEntryTimestamp;
+    if (!signature) {
+      throw new Error('no SET found in bundle');
+    }
+
+    if (!crypto.verifyBlob(data, publicKey, signature)) {
+      throw new Error('transparency log SET verification failed');
+    }
+  });
+}
+
+// Returns a properly formatted "VerificationPayload" for one of the
+// transaction log entires in the given bundle which can be used for SET
+// verification.
+function toVerificationPayload(bundle: Bundle, index = 0): VerificationPayload {
+  // Ensure bundle has tlog entries
+  const entries = bundle.verificationData?.tlogEntries;
+  if (!entries || entries.length - 1 < index) {
+    throw new Error('No tlog entries found in bundle');
+  }
+
+  // Rekor metadata
+  const { integratedTime, logIndex, logId } = entries[index];
+
+  if (!logId) {
+    throw new Error('No logId found in bundle');
+  }
+
+  // Recreate the Rekor entry from the bundle
+  const body = toVerificationBody(bundle);
+
+  return {
+    body: enc.base64Encode(json.canonicalize(body)),
+    integratedTime: Number(integratedTime),
+    logIndex: Number(logIndex),
+    logID: logId.keyId.toString('hex'),
+  };
+}
+
+// Recreates the original Rekor entry from the bundle.
+function toVerificationBody(bundle: Bundle): EntryKind {
+  // Extract the public key (or signing cert) from the bundle
+  if (!bundle.verificationMaterial) {
+    throw new Error('No verification material found in bundle');
+  }
+
+  const publicKey = toPublicKey(bundle.verificationMaterial);
+
+  switch (bundle.content?.$case) {
+    case 'messageSignature': {
+      return toMessageSignatureVerificationBody(
+        bundle.content.messageSignature,
+        publicKey
+      );
+    }
+    case 'dsseEnvelope': {
+      return toDSSEVerificationBody(bundle.content.dsseEnvelope, publicKey);
+    }
+    default:
+      throw new Error('Unsupported bundle type');
+  }
+}
+
+function toPublicKey(verificationMaterial: VerificationMaterial): string {
+  switch (verificationMaterial?.content?.$case) {
+    case 'x509CertificateChain': {
+      const der =
+        verificationMaterial.content.x509CertificateChain.certificates[0];
+      return pem.fromDER(der.derBytes);
+    }
+    case 'publicKey':
+    // TODO: How to handle this?
```

Comment:
Do we need to figure this out for verifying the SET of publish attestation?

Reply:
Yeah. I'll want to see an example of one of those entries in Rekor in order to figure out how this should work.

Reply:
Here's an example: https://rekor.sigstore.dev/api/v1/log/entries?logIndex=5983279
and here's the bundle: https://github.com/github/package-security/issues/439